Lib/Android/FreePIE Android IMU/src/com/freepie/android/imu/MainActivity.java

```diff
@@ -22,46 +30,95 @@
-
-	private UdpSenderTask udpSender;
-	private static final String IP = "ip";
-	private static final String PORT = "port";
-	private static final String INDEX = "index";
-	private static final String SEND_ORIENTATION = "send_orientation";
-	private static final String SEND_RAW = "send_raw";
-	private static final String SAMPLE_RATE = "sample_rate";
-	private static final String DEBUG = "debug";
-	
-	private static final String DEBUG_FORMAT = "%.2f;%.2f;%.2f";
-	
-	private ToggleButton start;
-	private EditText txtIp;
-	private EditText txtPort;
-	private Spinner spnIndex;
-	private CheckBox chkSendOrientation;
-	private CheckBox chkSendRaw;
-	private Spinner spnSampleRate;
-	private CheckBox chkDebug; 
-	private LinearLayout debugView;
-	private TextView acc;    
-	private TextView gyr;
-	private TextView mag;
-	private TextView imu;
-		
+public class MainActivity extends Activity {
+    private UdpSenderService udpSenderService;
+
+    private ServiceConnection conn = new ServiceConnection() {
+        @Override
+        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
+            udpSenderService = ((UdpSenderService.MyBinder) iBinder).getService();
+            start.setChecked(udpSenderService.isRunning());
+        }
+
+        @Override
+        public void onServiceDisconnected(ComponentName componentName) {
+            udpSenderService = null;
+        }
+    };
+
+    private void bindService() {
+        startService(new Intent(this, UdpSenderService.class));
+        bindService(new Intent(this, UdpSenderService.class), conn, Context.BIND_AUTO_CREATE);
+    }
+
+    private void unbindService() {
+        if (udpSenderService != null) {
+            unbindService(conn);
+        }
+    }
+
+    private static final String IP = "ip";
+    private static final String PORT = "port";
+    private static final String INDEX = "index";
+    private static final String SEND_ORIENTATION = "send_orientation";
+    private static final String SEND_RAW = "send_raw";
+    private static final String SAMPLE_RATE = "sample_rate";
+    private static final String DEBUG = "debug";
+
+    private static final String DEBUG_FORMAT = "%.2f;%.2f;%.2f";
+
+    private ToggleButton start;
+    private TextView lblIP;
+    private EditText txtIp;
+    private EditText txtPort;
+    private Spinner spnIndex;
+    private CheckBox chkSendOrientation;
+    private CheckBox chkSendRaw;
+    private Spinner spnSampleRate;
+    private CheckBox chkDebug;
+    private LinearLayout debugView;
+    private TextView acc;
+    private TextView gyr;
+    private TextView mag;
+    private TextView imu;
+    private LinearLayout emptyLayout;
+
+    private TimerTask debugHandler = new TimerTask() {
+        @Override
+        public void run() {
+            runOnUiThread(new Runnable() {
+                @Override
+                public void run() {
+                    if (udpSenderService != null && udpSenderService.isRunning()) {

```

Comment:
I made a mistake here. Please add checking <code>chkDebug.isChecked()</code> before allocating data with <code>new</code>. It runs on UI thread anyway so it can do with no locks.